File: client/lib/temples.ts
export type Level = 'Safe' | 'Warning' | 'Critical' | 'No data';
export type SourceType = 'snapshot' | 'mjpeg' | 'stream' | 'push' | 'file' | 'demo';

export interface Latest {
  ts: string;
  count: number;
  level: Level;
  source: string;
  people_per_m2?: number | null;
}

export interface Camera {
  id: number;
  temple_id: number;
  name: string;
  source_type: SourceType;
  url: string | null;
  area_m2: number | null;
  zone_capacity: number | null;
  interval_seconds: number;
  enabled: boolean;
  last_polled_at: string | null;
  last_status: string | null;
  last_error: string | null;
  latest: Latest | null;
  online: boolean;
  has_frame: boolean;
}

export interface Temple {
  id: number;
  name: string;
  deity: string | null;
  city: string | null;
  state: string | null;
  address: string | null;
  latitude: number | null;
  longitude: number | null;
  capacity: number;
  area_m2: number | null;
  warn: number;
  crit: number;
  opening_time: string | null;
  closing_time: string | null;
  contact_name: string | null;
  contact_phone: string | null;
  contact_email: string | null;
  notes: string | null;
  current_count: number | null;
  occupancy_pct: number | null;
  level: Level;
  cameras_total: number;
  cameras_online: number;
  last_update: string | null;
  color?: string;
  cameras?: Camera[];
}

export const LEVEL_CHIP: Record<string, string> = {
  Safe: 'bg-green-100 text-green-700 border-green-200 dark:bg-green-900/40 dark:text-green-300 dark:border-green-800',
  Warning: 'bg-amber-100 text-amber-700 border-amber-200 dark:bg-amber-900/40 dark:text-amber-300 dark:border-amber-800',
  Critical: 'bg-red-100 text-red-700 border-red-200 dark:bg-red-900/40 dark:text-red-300 dark:border-red-800',
  'No data': 'bg-gray-100 text-gray-600 border-gray-200 dark:bg-gray-800 dark:text-gray-300 dark:border-gray-700',
};

export const LEVEL_BAR: Record<string, string> = {
  Safe: 'bg-green-500',
  Warning: 'bg-amber-500',
  Critical: 'bg-red-500',
  'No data': 'bg-gray-400',
};

export const SOURCE_INFO: Record<SourceType, { label: string; help: string; placeholder: string }> = {
  snapshot: {
    label: 'HTTP snapshot (pull)',
    help: 'The server fetches one JPEG from this URL every interval. Works with most IP cameras (e.g. http://192.168.1.20/snapshot.jpg).',
    placeholder: 'http://192.168.1.20/snapshot.jpg',
  },
  mjpeg: {
    label: 'MJPEG stream (pull)',
    help: 'The server opens the motion-JPEG stream, takes one frame and closes it.',
    placeholder: 'http://192.168.1.20:8080/video',
  },
  stream: {
    label: 'RTSP / RTMP / HLS / webcam (pull, needs OpenCV)',
    help: 'Any stream OpenCV can open. Credentials in the URL are stored server-side and never shown again. A number such as 0 uses a webcam attached to the server.',
    placeholder: 'rtsp://user:pass@192.168.1.20:554/stream1',
  },
  push: {
    label: 'Push (camera / edge device sends frames)',
    help: 'No URL. After saving you get a private ingest URL; your camera, an ffmpeg loop or an edge script POSTs frames to it.',
    placeholder: '',
  },
  demo: {
    label: 'Demo feed (replays ShanghaiTech test images)',
    help: 'No hardware needed: the server replays test images through the real CSRNet pipeline on a daily crowd pattern. URL = pattern name (somnath, dwarka, rameswaram, nathdwara or leave empty).',
    placeholder: 'somnath',
  },
  file: {
    label: 'Video / image file on the server (demo)',
    help: 'For demos: a recorded video is replayed as if live. Only works if the server is started with ALLOW_FILE_CAMERAS=1.',
    placeholder: '/data/recordings/gate.mp4',
  },
};

export function timeAgo(iso: string | null): string {
  if (!iso) return 'never';
  const s = Math.max(0, (Date.now() - new Date(iso).getTime()) / 1000);
  if (s < 60) return `${Math.round(s)}s ago`;
  if (s < 3600) return `${Math.round(s / 60)} min ago`;
  if (s < 86400) return `${Math.round(s / 3600)} h ago`;
  return `${Math.round(s / 86400)} d ago`;
}

File: backend/routes/temples.py
from __future__ import annotations

import os
import secrets
import sqlite3
from datetime import datetime, timedelta, timezone
from typing import Literal

from fastapi import APIRouter, Depends, File, Header, HTTPException, Query, Response, UploadFile
from fastapi.responses import FileResponse
from pydantic import BaseModel, Field, model_validator

import db
from services.capture import FRAMES_DIR, capture_camera, load_camera, mark_error, process_frame
from services.frames import SOURCE_TYPES, validate_source
from services.status import COLORS, camera_view, latest_readings, temple_status

router = APIRouter(tags=["temples"])


def require_admin(x_api_key: str | None = Header(default=None)) -> None:
    """If ADMIN_API_KEY is set on the server, every management call must send it as X-API-Key."""
    expected = os.getenv("ADMIN_API_KEY")
    if expected and not secrets.compare_digest(x_api_key or "", expected):
        raise HTTPException(status_code=401, detail="Missing or wrong X-API-Key.")


admin = Depends(require_admin)


# ------------------------------------------------------------------ schemas
class TempleIn(BaseModel):
    name: str = Field(min_length=2, max_length=120)
    deity: str | None = Field(default=None, max_length=120)
    city: str | None = Field(default=None, max_length=80)
    state: str | None = Field(default=None, max_length=80)
    address: str | None = Field(default=None, max_length=300)
    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)
    capacity: int = Field(gt=0, description="Maximum safe number of people present at once")
    area_m2: float | None = Field(default=None, gt=0)
    warn: float | None = Field(default=None, gt=0, description="People at which the temple becomes Warning (default 70 % of capacity)")
    crit: float | None = Field(default=None, gt=0, description="People at which the temple becomes Critical (default 90 % of capacity)")
    opening_time: str | None = Field(default=None, max_length=10)
    closing_time: str | None = Field(default=None, max_length=10)
    contact_name: str | None = Field(default=None, max_length=120)
    contact_phone: str | None = Field(default=None, max_length=40)
    contact_email: str | None = Field(default=None, max_length=120)
    notes: str | None = Field(default=None, max_length=1000)

    @model_validator(mode="after")
    def _thresholds(self) -> "TempleIn":
        if (self.warn is None) != (self.crit is None):
            raise ValueError("Set both warn and crit, or leave both empty to use 70 % / 90 % of capacity.")
        if self.warn is not None and self.crit is not None and self.warn >= self.crit:
            raise ValueError("warn must be smaller than crit.")
        return self


class CameraIn(BaseModel):
    name: str = Field(min_length=2, max_length=120)
    source_type: Literal["snapshot", "mjpeg", "stream", "push", "file", "demo"]
    url: str | None = Field(default=None, max_length=600)
    area_m2: float | None = Field(default=None, gt=0, description="Ground area the camera covers (enables people/m2)")
    zone_capacity: int | None = Field(default=None, gt=0, description="Safe capacity of the zone in view (optional)")
    interval_seconds: int = Field(default=60, ge=5, le=3600)
    enabled: bool = True

    @model_validator(mode="after")
    def _source(self) -> "CameraIn":
        validate_source(self.source_type, self.url)
        return self


def _temple_values(t: TempleIn) -> dict:
    d = t.model_dump()
    if d["warn"] is None:
        d["warn"], d["crit"] = round(0.7 * t.capacity), round(0.9 * t.capacity)
    return d


def _get_temple(c: sqlite3.Connection, temple_id: int) -> dict:
    t = db.one(c.execute("SELECT * FROM temples WHERE id=?", (temple_id,)))
    if not t:
        raise HTTPException(status_code=404, detail="Temple not found.")
    return t


# ------------------------------------------------------------------ temples
@router.get("/temples")
def list_temples() -> dict:
    with db.connect() as c:
        temples = db.rows(c.execute("SELECT * FROM temples ORDER BY name"))
        items = [temple_status(c, t) for t in temples]
    for i, t in enumerate(items):
        t["color"] = COLORS[i % len(COLORS)]
    return {"temples": items}


@router.post("/temples", status_code=201, dependencies=[admin])
def create_temple(body: TempleIn) -> dict:
    v = _temple_values(body)
    cols = ", ".join(v)
    marks = ", ".join("?" for _ in v)
    try:
        with db.connect() as c:
            cur = c.execute(f"INSERT INTO temples ({cols}, created_at) VALUES ({marks}, ?)", (*v.values(), db.now_iso()))
            return temple_status(c, _get_temple(c, cur.lastrowid), with_cameras=True)
    except sqlite3.IntegrityError as exc:
        raise HTTPException(status_code=409, detail="A temple with this name already exists.") from exc


@router.get("/temples/{temple_id}")
def get_temple(temple_id: int) -> dict:
    with db.connect() as c:
        return temple_status(c, _get_temple(c, temple_id), with_cameras=True)


@router.put("/temples/{temple_id}", dependencies=[admin])
def update_temple(temple_id: int, body: TempleIn) -> dict:
    v = _temple_values(body)
    sets = ", ".join(f"{k}=?" for k in v)
    try:
        with db.connect() as c:
            _get_temple(c, temple_id)
            c.execute(f"UPDATE temples SET {sets} WHERE id=?", (*v.values(), temple_id))
            return temple_status(c, _get_temple(c, temple_id), with_cameras=True)
    except sqlite3.IntegrityError as exc:
        raise HTTPException(status_code=409, detail="A temple with this name already exists.") from exc


@router.delete("/temples/{temple_id}", status_code=204, response_class=Response, dependencies=[admin])
def delete_temple(temple_id: int):
    with db.connect() as c:
        _get_temple(c, temple_id)
        ids = [r["id"] for r in db.rows(c.execute("SELECT id FROM cameras WHERE temple_id=?", (temple_id,)))]
        c.execute("DELETE FROM temples WHERE id=?", (temple_id,))
    for cid in ids:
        (FRAMES_DIR / f"{cid}.jpg").unlink(missing_ok=True)


@router.get("/temples/{temple_id}/readings")
def temple_readings(temple_id: int, hours: int = Query(24, ge=1, le=720), bucket_minutes: int = Query(5, ge=1, le=240)) -> dict:
    since = (datetime.now(timezone.utc) - timedelta(hours=hours)).isoformat(timespec="seconds")
    with db.connect() as c:
        t = _get_temple(c, temple_id)
        raw = db.rows(c.execute("SELECT ts, camera_id, count FROM readings WHERE temple_id=? AND ts>=? ORDER BY ts", (temple_id, since)))
    buckets: dict[str, dict[int, float]] = {}
    for r in raw:
        dt = datetime.fromisoformat(r["ts"])
        floor = dt - timedelta(minutes=dt.minute % bucket_minutes, seconds=dt.second)
        buckets.setdefault(floor.isoformat(timespec="seconds"), {})[r["camera_id"]] = r["count"]  # last reading of the camera in the bucket
    points = [{"ts": k, "total": round(sum(v.values()), 1), "cameras": v} for k, v in sorted(buckets.items())]
    return {"warn": t["warn"], "crit": t["crit"], "capacity": t["capacity"], "points": points}


# ------------------------------------------------------------------ cameras
def _new_key() -> str:
    return secrets.token_urlsafe(24)


@router.post("/temples/{temple_id}/cameras", status_code=201, dependencies=[admin])
def add_camera(temple_id: int, body: CameraIn) -> dict:
    with db.connect() as c:
        _get_temple(c, temple_id)
        cur = c.execute(
            """INSERT INTO cameras(temple_id,name,source_type,url,area_m2,zone_capacity,interval_seconds,enabled,api_key,created_at)
               VALUES(?,?,?,?,?,?,?,?,?,?)""",
            (temple_id, body.name, body.source_type, body.url, body.area_m2, body.zone_capacity, body.interval_seconds,
             int(body.enabled), _new_key(), db.now_iso()),
        )
        cam = db.one(c.execute("SELECT * FROM cameras WHERE id=?", (cur.lastrowid,)))
    return {**camera_view(cam, None), "api_key": cam["api_key"]}


@router.put("/cameras/{camera_id}", dependencies=[admin])
def update_camera(camera_id: int, body: CameraIn) -> dict:
    with db.connect() as c:
        if not db.one(c.execute("SELECT id FROM cameras WHERE id=?", (camera_id,))):
            raise HTTPException(status_code=404, detail="Camera not found.")
        url = body.url
        if url and "***@" in url:  # the client echoed the masked url back: keep the stored credentials
            url = db.one(c.execute("SELECT url FROM cameras WHERE id=?", (camera_id,)))["url"]
        c.execute(
            "UPDATE cameras SET name=?, source_type=?, url=?, area_m2=?, zone_capacity=?, interval_seconds=?, enabled=? WHERE id=?",
            (body.name, body.source_type, url, body.area_m2, body.zone_capacity, body.interval_seconds, int(body.enabled), camera_id),
        )
        cam = db.one(c.execute("SELECT * FROM cameras WHERE id=?", (camera_id,)))
        latest = latest_readings(c, cam["temple_id"]).get(camera_id)
    return camera_view(cam, latest)


@router.delete("/cameras/{camera_id}", status_code=204, response_class=Response, dependencies=[admin])
def delete_camera(camera_id: int):
    with db.connect() as c:
        if not db.one(c.execute("SELECT id FROM cameras WHERE id=?", (camera_id,))):
            raise HTTPException(status_code=404, detail="Camera not found.")
        c.execute("DELETE FROM cameras WHERE id=?", (camera_id,))
    (FRAMES_DIR / f"{camera_id}.jpg").unlink(missing_ok=True)


@router.get("/cameras/{camera_id}/integration", dependencies=[admin])
def camera_integration(camera_id: int) -> dict:
    """Ingest URL and ready-to-paste snippets for pushing frames from a camera / edge device."""
    cam, _ = load_camera_or_404(camera_id)
    path = f"/ingest/{cam['api_key']}"
    return {
        "camera_id": camera_id,
        "source_type": cam["source_type"],
        "ingest_path": path,
        "curl": f'curl -X POST "$API_BASE{path}" -F "image=@frame.jpg"',
        "ffmpeg_loop": (
            "while true; do ffmpeg -loglevel error -y -rtsp_transport tcp -i \"rtsp://USER:PASS@CAMERA_IP/stream\" "
            f"-frames:v 1 frame.jpg && curl -s -X POST \"$API_BASE{path}\" -F \"image=@frame.jpg\"; sleep 30; done"
        ),
        "python": (
            "import requests\n"
            f"requests.post(API_BASE + '{path}', files={{'image': open('frame.jpg', 'rb')}}, timeout=30)"
        ),
    }


def load_camera_or_404(camera_id: int):
    try:
        return load_camera(camera_id)
    except LookupError as exc:
        raise HTTPException(status_code=404, detail="Camera not found.") from exc


@router.post("/cameras/{camera_id}/capture", dependencies=[admin])
def capture_now(camera_id: int, store: bool = Query(True, description="false = test the feed without saving a reading")) -> dict:
    cam, _ = load_camera_or_404(camera_id)
    if cam["source_type"] == "push":
        raise HTTPException(status_code=400, detail="Push cameras send their own frames to the ingest URL; nothing to pull.")
    try:
        return capture_camera(camera_id, store, randomize=True)  # manual click: random dataset image for demo cameras
    except Exception as exc:
        raise HTTPException(status_code=502, detail=f"Could not read the camera: {exc}") from exc


@router.post("/ingest/{api_key}")
async def ingest_frame(api_key: str, image: UploadFile = File(...)) -> dict:
    """Push endpoint for cameras / edge devices. The per-camera key in the URL is the credential."""
    with db.connect() as c:
        cam = db.one(c.execute("SELECT id, enabled FROM cameras WHERE api_key=?", (api_key,)))
    if not cam or not cam["enabled"]:
        raise HTTPException(status_code=404, detail="Unknown or disabled camera key.")
    data = await image.read()
    if not data:
        raise HTTPException(status_code=400, detail="Empty image.")
    try:
        r = process_frame(cam["id"], data)
    except ValueError as exc:
        mark_error(cam["id"], str(exc))
        raise HTTPException(status_code=400, detail="Not a valid image.") from exc
    return {"ts": r["ts"], "count": r["count"], "level": r["level"], "source": r["source"]}


@router.get("/cameras/{camera_id}/latest.jpg")
def latest_frame(camera_id: int) -> FileResponse:
    path = FRAMES_DIR / f"{camera_id}.jpg"
    if not path.exists():
        raise HTTPException(status_code=404, detail="No frame captured yet.")
    return FileResponse(path, media_type="image/jpeg", headers={"Cache-Control": "no-store"})
